data_pipeline/gmail_production_data/get_gmail_data.py

```diff
@@ -0,0 +1,141 @@
+    label_response = service.users().labels().list(userId=userId).execute()
+    df = pd.DataFrame(label_response['labels'])
+    return df
+
+
+def iter_link_message_thread(service, userId):
+    response = service.users().messages().list(userId=userId).execute()
+    if 'messages' in response:
+        yield from response['messages']
+    while ('nextPageToken' in response):
+        page_token = response['nextPageToken']
+        response = service.users().messages().list(userId=userId, pageToken=page_token).execute()
+        yield from response['messages']
+
+
+def get_link_message_thread(service, userId):
+    df = pd.DataFrame(iter_link_message_thread(service, userId))
+    return df
+
+
+@retry(Exception, tries=5)
+def get_one_thread(service, userId, threadId):
+    """
+    To get thread details like :
+        - threadId
+        - historyId
+        - total_email_count
+        - sent_email_count
+        - first_responder
+        - first_reponse_date
+    """
+    thread_results = service.users().threads().get(userId=userId, id=threadId).execute()
+
+    df = pd.DataFrame()
+
+    df['threadId'] = [thread_results['id']]
+    df['historyId'] = [thread_results['historyId']]
+    df['total_email_count'] = [len(thread_results['messages'])]
+
+    # to find count of emails from elifesciences.org in a thread (sent_email_count)
+    sent_email_count = 0
+    EMAIL_SENT_PATTERN = 'elifesciences.org'
+    for i in range(len(thread_results['messages'])):
+        header_list = thread_results['messages'][i]['payload']['headers']
+        for j in range(len(header_list)):
+            if header_list[j]['name'] == 'From':
+                if EMAIL_SENT_PATTERN in header_list[j]['value']:
+                    sent_email_count += 1
+
+    df['sent_email_count'] = sent_email_count
+
+    # to get first responder
+    FISRT_RESPONSE_ITERATION = 0
+    for i in range(len(thread_results['messages'])):
+        header_list = thread_results['messages'][i]['payload']['headers']
+        for j in range(len(header_list)):
+            if header_list[j]['name'] == 'From':
+                if 'elifesciences.org' in header_list[j]['value']:
+                    df['first_responder'] = [header_list[j]['value']]
+                    FISRT_RESPONSE_ITERATION = i
+                    break
+
+    # to get first response date
+    header_list = thread_results['messages'][FISRT_RESPONSE_ITERATION]['payload']['headers']
+    for k in range(len(header_list)):
+        if header_list[k]['name'] == 'Date':
+            df['first_reponse_date'] = [header_list[k]['value']]
+            break
+
+    return df
+
+
+def get_thread_details_for_given_ids(service, userId, threadIds_file):
+    df = pd.read_csv(threadIds_file, usecols=['threadId'])
+    threadIds_arr = df['threadId'].unique()
+    df = pd.DataFrame()
+    for threadId in threadIds_arr:
+        df_one_thread = get_one_thread(service, userId, threadId)
+        df = df.append(df_one_thread)
```

Comment:
[merge](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html) would probably be more efficient. Perhaps this is something where the data will get very large? Or not so large because you are doing the counting?